sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,364 @@
+            &program,
+            &(name.to_string() + "_texture_offset"),
+            Vector2::new(0, 0),
+        );
+
+        GlMovingWindowTexture {
+            id,
+            gl,
+            size,
+            u_texture_offset,
+            pixel_type: PhantomData,
+        }
+    }
+
+    /// Let's say the texture window has moved left and up. The new pixels that
+    /// are now inside the window are passed as a vertical and a horizontal
+    /// strip. Here they are in the texture window:
+    /// ```text
+    /// +-----+------------+
+    /// |#####|############| <- hori_strip
+    /// +-----|------------+
+    /// |#####|            |
+    /// |#####|            |
+    /// |#####|            |
+    /// +-----+------------+
+    ///    ^
+    ///    |
+    ///   vert_strip
+    /// ```
+    ///
+    /// You can see that they are redundant where they overlap. This is not
+    /// handled to keep the code from becoming more complex. Also, either may
+    /// be empty, which is the case for purely horizontal/vertical shifts.
+    ///
+    /// These strips have to be inserted correctly into the OpenGL texture,
+    /// which starts not at 0 but at `texture_offset` because of wrapping indexing.
+    /// Updating that texture from such a strip can typically not be done in one
+    /// step, but requires breaking it into up to four subregions. This is
+    /// because OpenGL textures can be read with wrapping indexing, but not
+    /// written with wrapping indexing, so we need to do it ourselves.
+    /// Let's look at the vertical strip:
+    ///
+    /// ```text
+    /// <- width ->
+    /// +---+-+
+    /// | 3 |4|
+    /// +---+-+
+    /// |   | |  vert_strip
+    /// | 1 |2|
+    /// |   | |
+    /// +---+-+
+    /// ```
+    ///
+    /// Wraparound at the right texture edge causes the split into regions 1/3
+    /// vs 2/4, and wraparound at the upper texture edge causes the split into
+    /// regions 1/2 vs 3/4.
+    /// Here's the texture with the vert_strip placed correctly:
+    ///
+    /// ```text
+    /// <- self.size ->
+    /// +-+-------+---+
+    /// | |       |   |
+    /// |2|       | 1 |
+    /// | |       |   |
+    /// +-+       +---+ <- texture_offset.y
+    /// |4|       | 3 |
+    /// +-+-------+---+
+    ///           ^
+    ///           |
+    ///           xoff
+    /// ```
+    ///
+    /// `xoff` is not equal to `texture_offset.x` if the texture moved right.
+    ///
+    /// The way this is handled is by creating regions as if they always need
+    /// splitting, and skipping empty regions that result from that.
+    pub fn incremental_update(
+        &mut self,
+        delta_x: i32,
+        delta_y: i32,
+        vert_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+        hori_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+    ) {
+        // width/height are always positive, but if the delta is negative, the
+        // start of the new region is actually offset + delta, not offset
```

Comment:
nit: let's add a comment what `xoff` and `yoff` are? I think there is a risk of confusion with `self.u_texture_offset.value.x`/`y`.

Reply:
Refactored that code and added comments.